### DISCUSSION TITLE
Cloud Support in the CLI

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#3677] [Pre-RFC]
### `edgedb cloud login`

Issues a browser-based authentication process, and store the access token in `<CONFIG_DIR>/cloud.json`.

### `edgedb cloud logout`

Clears out the stored access token in `<CONFIG_DIR>/cloud.json`.

## Changes in Existing Commands

A new instance type `cloud` will be added. It is a remote instance, but can be started/stopped/upgraded etc.

### Global parameter

A top-level option `--cloud-access-token` is available to override the stored access token in `<CONFIG_DIR>/cloud.json`. Meanwhile, this option is also available on the following commands for convenience:
 * `edgedb cloud`
 * `edgedb cloud login`
 * `edgedb instance create`
 * `edgedb instance list`
 * `edgedb instance start`
 * `edgedb instance stop`
 * `edgedb instance restart`
 * `edgedb instance destroy`
 * `edgedb instance link`
 * `edgedb instance logs`
 * `edgedb instance upgrade`
 * `edgedb instance revert`
 * `edgedb instance reset-password`
 * `edgedb project init`
 * `edgedb project unlink`
 * `edgedb project upgrade`

### `edgedb instance create`

This command will gain a new flag `--cloud`, when set, the new instance will be created in the default cloud organization.

TBD: `--org`, `--spec`

### `edgedb instance list`

New flag `--cloud`: list all cloud instances instead of local/remote instances.

### `edgedb instance start/stop/restart/logs/reset-password`

These commands will work on cloud instances in the same way as local instances.

### `edgedb instance destroy`

If the instance type is cloud, the data in the cloud will also be wiped out and the cloud instance will also be destroyed.

### `edgedb instance link`

This command will gain a new flag `--cloud`, when set, a differnt process/wizard will guide the user to link to an already-existing cloud instance for local access. The user will be asked for the name (and the org) of the cloud instance to link to, and the local name used in future `edgedb -I <INSTANCE_NAME>` access - the same name is proposed by default.

`--quiet`, `--non-interactive` options are also effective with `--cloud`. `--non-interactive` would require an explicit name given in the command line, which will be used for both cloud instance lookup and local reference.

TBD: `--org`

### `edgedb instance upgrade/revert`

TBD

### `edgedb project init`

This command will gain a new flag `--cloud`, when set, a differnt process/wizard will guide the user to initialize a/the project with a cloud instance.

Details TBD

### `edgedb project unlink`

If `--destroy-server-instance` is provided and the instance type is "cloud", the cloud instance will also be destroyed and the data will be lost.

### `edgedb project upgrade`

TBD

### 1. Add prefix/namespace to local reference names for cloud instances?

The current consents:

* One user could own/join multiple Organizations, organization names are globally unique
* Cloud instance has to belong to an organization, there will be a default organization ("personal"?)
* Cloud instance name is unique within an organization
* Organizations and instances can be renamed
* Organizations and instances all have immutable UUID identifiers

If we want to add prefix/namespace, the local reference name will likely look like `my_org/my_instance`.

The 1st question is more or less coupled with the 2nd question:

### 2. Should we automatically synchronize instance names in the cloud with local reference names?

For all options below, the cloud instance UUID is always the single source of truth for the linkage.

**Option 1.** Don't add prefix/namespace to local reference names, don't sync names. (*status quo*)

The organization name and instance name are only used in `edgedb instance link` to find the instance in the cloud to link to, the local reference name - though suggested by CLI automatically after the cloud names - can be totoally different and completely up to the user. Renaming in the cloud doesn't affect the local environment.

`edgedb instance create` will require a name from the user, the name will be used both in the cloud and as a local reference. If `--org my_org` is also specified, the local reference name will be like `my_org_my_instance`.

`edgedb project init` will suggest a name based on the project dir, and the user can modify it. Eventually the decided name will be used both in the cloud and as a local reference, regardless of the `--org` option.

`edgedb instance list --cloud` will show names in the cloud. If the instance is linked locally, the local reference name is also shown.

Pros:

 * Easiest to implement (~partially~ done for demo).

Cons:

 * Local and cloud names can be different and thus confusing.
 * If the cloud names conflict with local instance names, linking can be confusing.

**Option 2.** Prefix local reference names with org name, and automatically sync all renamings.

This option actually drops the concept of "local reference name" on the user side - you always use `org_name/instance_name` for cloud instances, and `instance_name` for pre-cloud local/remote instances. Synchronization happens automatically behind the scene.

Pros:
 * More intuitive to the users, because `my_org/my_instance` means the same instance anywhere.
 * No longer have to use the `--cloud` options (still available for explicitness).

Cons:

 * If renamed in the cloud, there might be a delay before the sync is done.
 * More work to do (inlcuding patches to the client bindings), complexity means more risks to break (but likely worth it!)

Implementation detail (proposal):

* Instead of treating cloud instances as a special remote instance, in this approach cloud instances should be handled separately. For example, there will be a new folder for cloud credentials JSON files.
* These cloud credentials JSON files are now considered cache. The CLI could name them in the most efficient way, and keep them updated in syncs.
* To reference the default organization, just use a leading slash, like `/my_instance`.
* Sync happens with a cloud-related command like `edgedb instance list --cloud`, and only sync if the last sync is older than a certain time.
* Newly-created cloud instances through the cloud web won't be immediately available to use in the client bindings through the `my_org/new_instance` name, a `edgedb instance link my_org/new_instance` is still necessary. (or not? At least in the CLI, we could automatically update the credentials JSON cache if the user is doing like `edgedb -I my_org/new_instance`).

**Option 3.** Like option 2 but less aggresive.

I don't think this is reasonable.

**Comment by tailhook** [comment]
> Issues a browser-based authentication process, and store the access token in <CONFIG_DIR>/cloud.json.

I don't think it makes sense to have a singleton. It should be `<CONFIG_DIR>/cloud-accounts/<account_id>.json` or similar.

We can have a "default" cloud account, though.

Additionally,

> A top-level option --cloud-access-token is available to override the stored access token in <CONFIG_DIR>/cloud.json. Meanwhile, this option is also available on the following commands for convenience:

While specifying token directly may make sense (although, I'm not sure what scope of the token could possible be). Specifying `<account_id>` would be helpful for many cases.

**Comment by tailhook** [comment]
> A new instance type cloud will be added. It is a remote instance, but can be started/stopped/upgraded etc.

What are mechanics of that? Is it a flag in credentials JSON? (I'm pretty sure it's a bad idea) Do we match credentials to cloud accounts somehow?

**Reply by fantix** [reply]
> What are mechanics of that? Is it a flag in credentials JSON?

For the prototype, it's the `cloud_instance_id` field in credentials JSON. For the proper implementation, I think most things related to the cloud should just be treated separately, like separate folder(s) for credentials JSON, separate handling branch logic, etc.

> Do we match credentials to cloud accounts somehow?

That's a very good question. In the prototype, if the user logs out and logs into another account, the already-linked cloud instances would only work partially (able to connect, unable to restart/destroy etc.). In this case, I think we should at least provide a more friendly error message like `cannot restart because this instance belongs to another cloud account`.

**Comment by tailhook** [comment]
> A top-level option --cloud-access-token is available to override the stored access token in <CONFIG_DIR>/cloud.json. Meanwhile, this option is also available on the following commands for convenience:

Also env variable should be available for the same purpose (to use in CI and other automation).

**Comment by tailhook** [comment]
Overall I'm +1 the option 1:

1. This will keep clients simple: only one credentials format is there.
2. There should be an alternative to `instance link` which is to get a credentials JSON from Web UI and upload it directly to the platform's secret store.
3. I'm generally against client bindings ever reading cloud credentials (which would be a natural extension for the option 2, to support effortless `account/instance` connection). So better prevent this from ever happening.

**Reply by elprans** [reply]
> I'm generally against client bindings ever reading cloud credentials

Any specific reason why?  From DX perspective it's attractive: you only need the API token and nothing else.  Otherwise to deploy you will need _both_ an API token and the ability to run `edgedb instance link`.  The downside is obviously that all bindings will need to implement the API.

**Reply by tailhook** [reply]
From local workflow it's quite minor. The reason for that is you need `project init` anyways, and we expect this as the best workflow for now.

But there are other DX issues:
1. Cloud tokens might be inconsistently supported in third-party bindings (as requires more work). This might be even more true as we add more features to the cloud: more fine-grained tokens, org accounts, whatever.
2. You will be unable to just pick credentials file and upload it to the hosting. Sometimes there will be credentials file and sometimes there will be a cloud token file and some other config which is project info or environment or whenever that `account/name` is.
3. Some people will try to put cloud token to production. I don't think this is useful even if we do properly scoped tokens, because why introduce another way to solve the same problem?
4. If you got used to API token switching back to local instance or credentials file requires learning something (scenario: you always used cloud, but now want to do something on an airplane)

**Reply by elprans** [reply]
> Cloud tokens might be inconsistently supported in third-party bindings (as requires more work).

Our bindings are already far from trivial, this wouldn't add too much complexity comparatively speaking, because it's a simple HTTP request and then the flow is the same as for the credentials file.

> You will be unable to just pick credentials file and upload it to the hosting.

Hm, why?  We can still allow you to download the full credentials file if you really need it.

> Some people will try to put cloud token to production.

We will encourage properly-scoped tokens by encoding instance names into tokens and adding "Get instance token" button to the dashboard (and an equivalent CLI command).

> If you got used to API token switching back to local instance or credentials file requires learning something

Quite the opposite, IMO.  `-I org/instance` will feel just as automagic as `-I local_instance`: no need to bother with DSNs or auth.

**Reply by tailhook** [reply]
> > Cloud tokens might be inconsistently supported in third-party bindings (as requires more work).
> 
> Our bindings are already far from trivial, this wouldn't add too much complexity comparatively speaking, because it's a simple HTTP request and then the flow is the same as for the credentials file.
> 

Not all languages have built-in HTTP. In 2022 HTTPS is quite complicated protocol, so this is a huge dependency at least in some languages (e.g. in Rust, and actually in (async) Python too)

> > You will be unable to just pick credentials file and upload it to the hosting.
> 
> Hm, why? We can still allow you to download the full credentials file if you really need it.
> 

It's not in your credentials directory, you have to learn how to do something to make that file.

> > Some people will try to put cloud token to production.
> 
> We will encourage properly-scoped tokens by encoding instance names into tokens and adding "Get instance token" button to the dashboard (and an equivalent CLI command).
> 

I don't understand how you see this. My understanding is that you do:
```
edgedb cloud authenticate
login: xxx
password: yyy
```
Now all your orgs are accessible with `-I someorg/instance`.

If to access `someorg/instance` you have to authenticate again and get another "properly scoped token", it isn't any better than what I'm proposing.

> > If you got used to API token switching back to local instance or credentials file requires learning something
> 
> Quite the opposite, IMO. `-I org/instance` will feel just as automagic as `-I local_instance`: no need to bother with DSNs or auth.

This is not what I mean. Think of every tutorial, and every deployment guide for a project must have a branch on if you're deploying to cloud do this, if you're deploying to your own instance do this. Rather then just always use credentials JSON got from the single CLI command (this is probably still simpler even if JSON file comes from different sources).

*(yes mentioning `local` maybe not the best thing here)*
